examples/core/3d-texture/app.js

```diff
@@ -0,0 +1,179 @@
+precision highp float;
+precision lowp sampler3D;
+in vec3 vUV;
+uniform sampler3D uTexture;
+uniform float uTime;
+out vec4 fragColor;
+void main() {
+  float alpha = texture(uTexture, vUV + vec3(0.0, 0.0, uTime)).r * 0.1;
+  fragColor = vec4(fract(vUV) * alpha, alpha);
+}`;
+
+const NEAR = 0.1;
+const FAR = 10.0;
+
+class AppAnimationLoop extends AnimationLoop {
+  constructor() {
+    super({debug: true});
+  }
+
+  getInfo() {
+    return INFO_HTML;
+  }
+
+  onInitialize({gl}) {
+    const perlin = noise3d.createPerlin({
+      interpolation: noise3d.interpolation.linear,
+      permutation: noise3d.array.shuffle(noise3d.array.range(0, 255), Math.random)
+    });
+
+    setParameters(gl, {
+      clearColor: [0, 0, 0, 1],
+      blend: true,
+      blendFunc: [gl.ONE, gl.ONE_MINUS_SRC_ALPHA]
+    });
+
+    // CREATE POINT CLOUD
+    const DIMENSIONS = 128;
+    const INCREMENT = 1 / DIMENSIONS;
+
+    const positionData = new Float32Array(DIMENSIONS * DIMENSIONS * DIMENSIONS * 3);
+    let positionIndex = 0;
+    let x = -0.5;
+    for (let i = 0; i < DIMENSIONS; ++i) {
+      let y = -0.5;
+      for (let j = 0; j < DIMENSIONS; ++j) {
+        let z = -0.5;
+        for (let k = 0; k < DIMENSIONS; ++k) {
+          positionData[positionIndex++] = x;
+          positionData[positionIndex++] = y;
+          positionData[positionIndex++] = z;
+          z += INCREMENT;
+        }
+        y += INCREMENT;
+      }
+      x += INCREMENT;
+    }
+
+    const positionBuffer = new Buffer(gl, positionData);
+
+    // CREATE 3D TEXTURE
+    const TEXTURE_DIMENSIONS = 16;
+    const NOISE_DIMENSIONS = TEXTURE_DIMENSIONS * 0.07;
+    const textureData = new Uint8Array(
+      TEXTURE_DIMENSIONS * TEXTURE_DIMENSIONS * TEXTURE_DIMENSIONS
+    );
+    let textureIndex = 0;
+    for (let i = 0; i < TEXTURE_DIMENSIONS; ++i) {
+      for (let j = 0; j < TEXTURE_DIMENSIONS; ++j) {
+        for (let k = 0; k < TEXTURE_DIMENSIONS; ++k) {
+          textureData[textureIndex++] =
+            (0.5 + 0.5 * perlin(i / NOISE_DIMENSIONS, j / NOISE_DIMENSIONS, k / NOISE_DIMENSIONS)) *
+            255;
+        }
+      }
+    }
+
+    const mvpMat = new Matrix4();
+    const viewMat = new Matrix4().lookAt({eye: [1, 1, 1]});
+
+    const texture = new Texture3D(gl, {
+      width: TEXTURE_DIMENSIONS,
+      height: TEXTURE_DIMENSIONS,
+      depth: TEXTURE_DIMENSIONS,
+      pixels: textureData,
+      format: gl.RED,
+      dataFormat: gl.R8
+    });
+
+    const cloud = new Model(gl, {
+      vs,
+      fs,
+      drawMode: gl.POINTS,
+      vertexCount: positionData.length / 3,
+      attributes: {
+        position: positionBuffer
+      },
+      uniforms: {
+        uTexture: texture,
+        uView: viewMat
+      }
+    });
+
+    const statsWidget = new StatsWidget(this.stats, {
+      containerStyle: 'position: absolute;top: 20px;left: 20px;'
+    });
+    statsWidget.setFormatter('CPU Time', stat => `CPU Time: ${stat.getAverageTime().toFixed(2)}ms`);
```

Comment:
Are we making these formatters defaults and moving them back into the widget? I feel this adds too much code to short demo apps...

Reply:
Yes, once I get this update out, I can clean up all the stats tracking code: https://github.com/uber-web/probe.gl/pull/64